### ISSUE TITLE
Error reading file opening form objects or vcx class objects

### ISSUE THREAD

**Issue by matthewaceware** [VFPX/GoFish#221]
## ℹ Computer information

- VFP version: 9.00.0000.5815
- GoFish version: 7.0.27

## 📝 Provide detailed reproduction steps (if any)

1. Search for something that lies in a form object.
2. Double-click to open the line of code.
3. Form opens correctly. The object opens correctly. You just get an error. 
![image](https://github.com/VFPX/GoFish/assets/31904139/522a732d-8e2a-44e7-8cd0-13c482b87433)
This does not seem to happen with form methods, i.e. Init. Prgs also open without issue.

### ✔️ Expected result

No error.

### ❌ Actual result

Error.

## 📷 Debugging Info
I've deleted my foxuser files and allowed the IDE to regenerate them. So, these are fresh files and I am still getting the same error. This has been happening in the last few GoFish builds, but never before that. No changes to my environment other than a Windows Update on Tuesday. No other IDE/Thor issues I've ran into other than what was previously reported and fixed.

**Comment by Jimrnelson** [comment]
From your description, this error is only occurring in an area of code that has just been released (selecting the control for the object), so the fact that it did not appear in earlier versions is not relevant.

This is a messy one to track down as I am not able to reproduce it AND this error is not a hard error that stops so that I can tell where it is occuring.

**Comment by Jimrnelson** [comment]
@matthewaceware 

Matthew --

I will need some assistance form you on this since I have no idea what line of code is causing this problem.

If you could do the following:

1. Run a GF search and navigate in the grid to a row that demonstrates the problem.
2. Get to the command window (Ctrl+F2)
3. Start capturing coverage information:    `Set Coverage to GF_Coverage.txt`
4. Get back to the GF grid row
5. double-click
6. When the annoying "Error reading FoxUser" message appears, count slowly to 12.  I want you to be patient.  (The elapsed time will be captured in the coverage log)
7. Finally click on OK
8. Return to command window
9. End capturing coverage information:    `Set Coverage to `
10. Open file explorer to get the TXT file justcreated.
11. Drag it from file explorer into a return message it (causing it to be attached.)

The hope is that will identify for me what line in what method is executing when you get the message.

**Comment by matthewaceware** [comment]
[gf_coverage.zip](https://github.com/VFPX/GoFish/files/15030702/gf_coverage.zip)
I looked at that file. I hope there is something there for you. I did notice a lot of references to FoxTabs. So, I shut that down, loaded GoFish fresh, and launched a line. That threw a different error: Error reading file 00000001: Incorrect function. 
![image](https://github.com/VFPX/GoFish/assets/31904139/fa923b12-d6a2-474d-87a2-fc882c098045)
So, I reran the Set Coverage to a different file for that:
[gf_coverage2.zip](https://github.com/VFPX/GoFish/files/15030751/gf_coverage2.zip)

Does that help?

Thanks,
Matthew

**Comment by Jimrnelson** [comment]
Did you wait for a looong time once the error message came up before clicking OK?  

10 seconds or longer?  I don't see any evidence of any delay.

That's what I am looking for.  the statement in the coverage that took 10 seconds or more (rather than a small fraction).

I got nothing if that line cannot be identified.

**Comment by Jimrnelson** [comment]
@matthewaceware 

Just reread your post and saw the reference to FoxTabs.  

I was never able to get FoxTabs to work reliably for me.  Eventually it would crash my session, so I gave up on it.

Can you try a FoxPro session where FoxTabs has never been run?  That "Error Reading ... Incorrect function" is a completely unfamiliar message.

**Comment by matthewaceware** [comment]
[gf_coverage3.zip](https://github.com/VFPX/GoFish/files/15031131/gf_coverage3.zip)
Yes, I am waiting at least 10 seconds. Above is from a session where foxtabs never ran.

**Comment by Jimrnelson** [comment]
@matthewaceware 

How disappointing.  There's no useful information there.

BTW, this time, which error did you get?  Foxuser or the Incorrect Function.  (Just curios, not really thinking it matters.)

For the time being, at least, I am out of ideas.

**Comment by klepetox** [comment]
I have reported the same error elsewhere with the test form, unfortunately no one has tried it. Now I know how the error occurs. I believe the fix will be easy. I am uploading the form again, the tooltips on the buttons describe the reason why the error occurs.

[test.zip](https://github.com/VFPX/GoFish/files/15033824/test.zip)

**Comment by Jimrnelson** [comment]
> I believe the fix will be easy.

Why do you believe the fix to be and why do you think it is easy?

**Comment by klepetox** [comment]
I mean, if it's clear how the bug is happening and the cause is very simple, it shouldn't be a problem to fix it. I'm sure you've fixed much more complicated and complex bugs.

**Comment by Jimrnelson** [comment]
> I mean, if it's clear how the bug is happening and the cause is very simple, it shouldn't be a problem to fix it. I'm sure you've fixed much more complicated and complex bugs.

I certainly appreciate your confidence in me but this is a very elusive bug.  Sometimes the most innocuous bugs are the hardest to eradicate.  In some rare cases, they actually indicate a problem in VFP itslef.

Until it's fixed, it ain't easy.

**Comment by Jimrnelson** [comment]
@matthewaceware @klepetox @DougHennig 

First of all, thank you @klepetox for providing the simplest possible code for reproducing the problem.  It fails for me as well.

Unfortunately, for the time being, I am marking this issue as "Can't fix".  

The error does indeed occur for me if I run the code normally but does NOT occur if I step through the debugger.  Thus I have no idea which line of code is causing the problem.  

Note that I said "for the time being".  I will circle back to this sometime in the hopes that a possible solution will occur to me.

**Comment by klepetox** [comment]
The oldest version of GF I have is 7.0.18 from March 23. The error is there already. This is even before the changes for selecting  controls were made, or not?

**Comment by Jimrnelson** [comment]
@klepetox, @matthewaceware 

> The oldest version of GF I have is 7.0.18 from March 23. The error is there already. This is even before the changes for selecting controls were made, or not?

This is a **very** helpful comment, as it steered my attention in a different direction.

I have verified that this problem occurs in code from GF6 and the error occurs after all the code for the dbl-click has executed and focus transfers to the code window.

Now, I have found one distinction between the code in the two command buttons:  Cmd1 has a trailing CR, Cmd2 does not.

I can replicate the problem in Cmd1 by removing the trailing CR and can remove the problem for Cmd2 by adding the CR.

It appears, without any further testing, that the failure occurs if the match line is the last line of code and there are no trailing CRs.

Can you verify that this is what is happening?  (This is correctable)

This might be all that is going on.

**Comment by Jimrnelson** [comment]
OMG, it would have been so much more helpful if you had put that in the message here as well.  It would have saved me quite some time.  It never occurred to me to actually RUN the form.

In any case, a followup --Has it always been the case for you that the problem occurs on the last line of a method and that line does not have a trailing CR?  (It would be marvelous if it were that simple!)

**Comment by klepetox** [comment]
Yes, it seems so.

`WAIT WINDOW 1
GO bottom
WAIT WINDOW 2`
- no error

`WAIT WINDOW 1
GO bottom`
- error

`GO bottom`
- error

`GO bottom
`
- no error

**Comment by klepetox** [comment]
Yes, it seems so. Only last row without CRLF.

**Comment by Jimrnelson** [comment]
@klepetox @matthewaceware 

I believe that this bug has been taken care of in today's update 7.0.27B.

Hats off to @klepetox on this one, as he supplied everything I needed to narrow the issue to one line of code:

1. He identified that the error preceded the release of the update to select the control for a method (#194).  This narrowed the search to a single method (implemented in GF6).
2. He identified that the error did not occur in GF6.  This narrowed the search in that method to the few lines I had added/modified.
3. He identified that the error occurred only if the match line was the last line of code with no trailing CR.

This bug can be reproduced (not easily, though) as it's a problem with _EdGetStr if the third parameter is greater than the number of characters in the entire method.

Here's hoping that we can put this one to bed.
